### PR TITLE
RT-108 handle multiple mime parts

### PR DESCRIPTION
**Pull request by ismail-s**
- Use [`defusedxml`](https://pypi.org/project/defusedxml/) in more places in the codebase for parsing xml, as it protects us against various XML vulnerabilities that the Python stdlib doesn't
- Stop using legacy methods of the `email` Python stdlib module (ie use [`EmailMessage`](https://docs.python.org/3.7/library/email.message.html#email.message.EmailMessage) and only the methods defined on it, and none of the methods defined on [`Message`](https://docs.python.org/3.7/library/email.compat32-message.html#email.message.Message)). Also stop using the legacy [`compat32`](https://docs.python.org/3.7/library/email.policy.html#email.policy.compat32) policy.
- Handle serialisation/deserialisation of optional additional attachments in `ebxml_request_envelope.py`

### CODE REVIEW

File: mhs/common/mhs_common/messages/ebxml_request_envelope.py [hunk 1]
```diff
@@ -54,19 +113,39 @@ def from_string(cls, headers: Dict[str, str], message: str) -> EbxmlRequestEnvel
...
+    def _add_descriptions_to_attachments(cls, xml_tree, attachments):
+        for attachment in attachments:
+            description_element = xml_tree.find(
+                f".//{ebxml_envelope.EBXML_NAMESPACE}:Reference["
+                f"@{ebxml_envelope.XLINK_NAMESPACE}:href='cid:{attachment[ATTACHMENT_CONTENT_ID]}']"
+                f"/{ebxml_envelope.EBXML_NAMESPACE}:Description",
+                ebxml_envelope.NAMESPACES)
+            if description_element is None:
+                logger.info('0003', "{Attachment} with {ContentType} found with no description. "
+                                    "Setting description to ''",
```
Comment (ismail-s) [inline mhs/common/mhs_common/messages/ebxml_request_envelope.py:157]: Around the description, I've made it mandatory when serialising, and when deserialising, as it's optional in the EIS (see links to EIS on the [JIRA story](https://gpitbjss.atlassian.net/browse/RT-107)), I've put this code here to handle it being optional by setting it as `''`.

@colin-bateson do you have any thoughts about this, do you think I should do things differently?

File: mhs/common/mhs_common/messages/ebxml_request_envelope.py [hunk 2]
```diff
@@ -26,22 +29,78 @@
...
+            }
         """
         super().__init__(EBXML_TEMPLATE, message_dictionary)
 
-    def serialize(self) -> Tuple[str, Dict[str, str], str]:
-        message_id, http_headers, message = super().serialize()
+    def serialize(self, message_dictionary=None) -> Tuple[str, Dict[str, str], str]:
+        message_dictionary = copy.deepcopy(self.message_dictionary)
+
+        attachment: dict
```
Comment (ismail-s) [inline mhs/common/mhs_common/messages/ebxml_request_envelope.py:100]: TODO-extract serialisation of attachments into method

File: mhs/common/mhs_common/messages/ebxml_request_envelope.py [hunk 3]
```diff
@@ -54,19 +113,39 @@ def from_string(cls, headers: Dict[str, str], message: str) -> EbxmlRequestEnvel
...
+
+        logger.info('0002', 'Extracted {extracted_values} from message', {'extracted_values': extracted_values})
 
         if payload_part:
             extracted_values[MESSAGE] = payload_part
 
         return EbxmlRequestEnvelope(extracted_values)
 
+    @classmethod
+    def _add_descriptions_to_attachments(cls, xml_tree, attachments):
```
Comment (ismail-s) [inline mhs/common/mhs_common/messages/ebxml_request_envelope.py]: TODO-bung docstring

File: mhs/common/data/templates/ebxml_request.mustache [hunk 4]
```diff
@@ -49,4 +54,12 @@ Content-Type: text/xml; charset=UTF-8
 Content-Transfer-Encoding: 8bit
 
 {{{hl7_message}}}
+{{#attachments}}
+----=_MIME-Boundary
+Content-Id: <{{content_id}}>
```
Comment (ismail-s) [inline mhs/common/data/templates/ebxml_request.mustache:59]: TODO-double-check whether this needs to end with @some-url-here

File: mhs/common/mhs_common/messages/ebxml_request_envelope.py [hunk 5]
```diff
@@ -107,19 +186,67 @@ def _extract_message_parts(msg: email.message.Message) -> Tuple[str, str]:
...
         payload_part = None
+        attachments = []
         if len(message_parts) > 1:
-            payload_part = message_parts[1].get_payload()
+            payload_part, is_base64_payload = EbxmlRequestEnvelope._convert_message_part_to_str(message_parts[1])
+            if is_base64_payload:
+                logger.error('0008', 'Failed to decode HL7 payload part of message as text')
+                raise ebxml_envelope.EbXmlParsingError("Failed to decode HL7 payload part of message as text")
+
+            for attachment_message in message_parts[2:]:
```
Comment (ismail-s) [inline mhs/common/mhs_common/messages/ebxml_request_envelope.py]: TODO-maybe bung a comment above this to make it a bit clearer as to the flow of this function, maybe throughout put some comments. Or do as a separate method.

File: mhs/common/mhs_common/messages/ebxml_request_envelope.py [hunk 6]
```diff
@@ -107,19 +186,67 @@ def _extract_message_parts(msg: email.message.Message) -> Tuple[str, str]:
...
+                payload, is_base64 = EbxmlRequestEnvelope._convert_message_part_to_str(attachment_message)
+                attachment = {
+                    ATTACHMENT_PAYLOAD: payload,
+                    ATTACHMENT_BASE64: is_base64,
+                    ATTACHMENT_CONTENT_ID: str(attachment_message['Content-Id'][1:-1]),
```
Comment (ismail-s) [inline mhs/common/mhs_common/messages/ebxml_request_envelope.py]: TODO-bung comment here about < > removing

File: mhs/common/mhs_common/messages/tests/test_ebxml_request_envelope.py [hunk 7]
```diff
@@ -114,7 +176,33 @@ def test_serialize_required_tags(self, mock_get_uuid, mock_get_timestamp):
 
     @patch.object(message_utilities.MessageUtilities, "get_timestamp")
     @patch.object(message_utilities.MessageUtilities, "get_uuid")
-    def test_serialize_optional_tags(self, mock_get_uuid, mock_get_timestamp):
+    def test_serialize_attachment_required_tags(self, mock_get_uuid, mock_get_timestamp):
```
Comment (ismail-s) [inline mhs/common/mhs_common/messages/tests/test_ebxml_request_envelope.py]: TODO-describe behaviour of test better here

File: mhs/common/mhs_common/messages/tests/test_ebxml_request_envelope.py [hunk 8]
```diff
@@ -114,7 +176,33 @@ def test_serialize_required_tags(self, mock_get_uuid, mock_get_timestamp):
...
+                }]
+                del message_dictionary[ebxml_request_envelope.ATTACHMENTS][0][required_tag]
+                envelope = ebxml_request_envelope.EbxmlRequestEnvelope(message_dictionary)
+
+                with self.assertRaisesRegex(pystache_message_builder.MessageGenerationError, 'Failed to find key'):
+                    envelope.serialize()
+
+    @patch.object(message_utilities.MessageUtilities, "get_timestamp")
+    @patch.object(message_utilities.MessageUtilities, "get_uuid")
+    def test_serialize_boolean_tag_set_to_false(self, mock_get_uuid, mock_get_timestamp):
```
Comment (ismail-s) [inline mhs/common/mhs_common/messages/tests/test_ebxml_request_envelope.py]: TODO-improve test name

File: mhs/common/mhs_common/messages/tests/test_ebxml_request_envelope.py [hunk 9]
```diff
@@ -222,3 +372,29 @@ def test_from_string_missing_additional_values(self):
             with self.assertRaisesRegex(
                     ebxml_envelope.EbXmlParsingError, "Weren't able to find required attribute actor"):
                 ebxml_request_envelope.EbxmlRequestEnvelope.from_string(MULTIPART_MIME_HEADERS, message)
+
+        with self.subTest("A valid request containing one attachment without a description"):
```
Comment (ismail-s) [inline mhs/common/mhs_common/messages/tests/test_ebxml_request_envelope.py]: TODO-improve this description, mention empty string

File: mhs/common/mhs_common/messages/tests/test_ebxml_request_envelope.py [hunk 10]
```diff
@@ -222,3 +372,29 @@ def test_from_string_missing_additional_values(self):
...
+
+            parsed_message = ebxml_request_envelope.EbxmlRequestEnvelope.from_string(MULTIPART_MIME_HEADERS, message)
+
+            self.assertEqual(expected_values_with_payload, parsed_message.message_dictionary)
+
+    #######################
+    # Helper methods
+    #######################
+
+    def get_expected_file_string(self, filename: str):
```
Comment (ismail-s) [inline mhs/common/mhs_common/messages/tests/test_ebxml_request_envelope.py]: TODO-make this _

File: mhs/common/mhs_common/messages/tests/test_messages/ebxml_request_multiple_attachments.msg [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+Content-Transfer-Encoding: 8bit
+
+Some payload
+----=_MIME-Boundary
+Content-Id: <64A73E03-30BD-4231-9959-0C4B54400345>
+Content-Type: image/png
+Content-Transfer-Encoding: base64
+
+iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII=
+----=_MIME-Boundary--
```
Comment (ismail-s) [inline mhs/common/mhs_common/messages/tests/test_messages/ebxml_request_multiple_attachments.msg]: TODO-newlines for all of these files

File: mhs/common/mhs_common/messages/ebxml_envelope.py [hunk 12]
```diff
@@ -51,14 +54,16 @@ class _ElementToExtractWhenParsing(NamedTuple):
     def __init__(self, template_file: str, message_dictionary: Dict[str, Any]):
         super().__init__(template_file, message_dictionary)
 
-    def serialize(self) -> Tuple[str, Dict[str, str], str]:
+    def serialize(self, message_dictionary: Dict[str, Any] = None) -> Tuple[str, Dict[str, str], str]:
```
Comment (ismail-s) [inline mhs/common/mhs_common/messages/ebxml_envelope.py]: TODO-make this param _

**Review by ismail-s — COMMENTED** [review]
Review comments from over-the-shoulder review with @Philip-Woods

File: mhs/common/data/templates/ebxml_request.mustache (hunk 4 above)
Comment (ismail-s) [reply mhs/common/data/templates/ebxml_request.mustache:59]: After looking at EIS appendix A1.1.4 and [RFC 2392](https://tools.ietf.org/html/rfc2392#section-2), I've made the content ID be a url instead (82a220e).

**Review by Philip-Woods — APPROVED** [review]

**Pull request merged**